File: src/test/java/voidchess/player/ki/evaluation/EvaluatedMoveTest.java
package voidchess.player.ki.evaluation;

import org.testng.annotations.Test;
import voidchess.helper.Move;

import java.util.Iterator;
import java.util.NavigableSet;
import java.util.TreeSet;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;

public class EvaluatedMoveTest {
    @Test
    public void testGetter() {
        final Move move = Move.byCode("a2-a3");
        final double primaryValue = -2.91;
        final double secondaryValue = 1.2;

        EvaluatedMove eMove = new EvaluatedMove(move, new Ongoing(primaryValue, secondaryValue));

        assertEquals(eMove.getMove(), move);
        assertEquals(((Ongoing) (eMove.getValue())).getCombinedEvaluation(), primaryValue+secondaryValue);
    }

    @Test
    public void testCompareTo() {
        EvaluatedMove e1_1 = new EvaluatedMove(Move.byCode("b2-b4"), new Ongoing(.1f));
        EvaluatedMove e1_2 = new EvaluatedMove(Move.byCode("c2-c4"), new Ongoing(.1f));
        EvaluatedMove e2 = new EvaluatedMove(Move.byCode("b2-b4"), new Ongoing(-.1f));
        EvaluatedMove e3 = new EvaluatedMove(Move.byCode("f2-f4"), new Ongoing(1.2f));

        assertTrue(e1_1.compareTo(e1_1) == 0);
        assertTrue(e1_1.compareTo(e1_2) == 0);
        assertTrue(e1_1.compareTo(e2) > 0);
        assertTrue(e1_1.compareTo(e3) < 0);
    }

    @Test
    public void testOrderingInSortedSet() {

        NavigableSet<EvaluatedMove> set = new TreeSet<>();

        set.add(new EvaluatedMove(Move.byCode("a2-a3"), new Ongoing(-1.2f)));//7.
        set.add(new EvaluatedMove(Move.byCode("b2-b3"), new Ongoing(-0.5f)));//6.
        set.add(new EvaluatedMove(Move.byCode("c2-c3"), new Ongoing(4.6f)));//3.
        set.add(new EvaluatedMove(Move.byCode("d2-d3"), new Ongoing(2.1f)));//4.
        set.add(new EvaluatedMove(Move.byCode("e2-e3"), Draw.INSTANCE));//5.
        set.add(new EvaluatedMove(Move.byCode("f2-f3"), new CheckmateSelf(1)));//9.
        set.add(new EvaluatedMove(Move.byCode("g2-g3"), new CheckmateSelf(2)));//8.
        set.add(new EvaluatedMove(Move.byCode("h2-h3"), new CheckmateOther(1)));//1.
        set.add(new EvaluatedMove(Move.byCode("h2-h4"), new CheckmateOther(2)));//2.

        Iterator<EvaluatedMove> iter = set.descendingIterator();

        assertEquals(iter.next().getMove(), Move.byCode("h2-h3"));
        assertEquals(iter.next().getMove(), Move.byCode("h2-h4"));
        assertEquals(iter.next().getMove(), Move.byCode("c2-c3"));
        assertEquals(iter.next().getMove(), Move.byCode("d2-d3"));
        assertEquals(iter.next().getMove(), Move.byCode("e2-e3"));
        assertEquals(iter.next().getMove(), Move.byCode("b2-b3"));
        assertEquals(iter.next().getMove(), Move.byCode("a2-a3"));
        assertEquals(iter.next().getMove(), Move.byCode("g2-g3"));
        assertEquals(iter.next().getMove(), Move.byCode("f2-f3"));
    }
}

File: src/test/kotlin/voidchess/player/ki/evaluation/EvaluatedMoveTest.kt
package voidchess.player.ki.evaluation

import org.testng.annotations.Test
import voidchess.board.move.Move
import java.util.TreeSet

import org.testng.Assert.assertEquals
import org.testng.Assert.assertTrue

class EvaluatedMoveTest {
    @Test
    fun testGetter() {
        val move = Move.byCode("a2-a3")
        val primaryValue = -2.91
        val secondaryValue = 1.2

        val eMove = EvaluatedMove(move, Ongoing(primaryValue, secondaryValue))

        assertEquals(eMove.move, move)
        assertEquals((eMove.value as Ongoing).getCombinedEvaluation(), primaryValue + secondaryValue)
    }

    @Test
    fun testCompareTo() {
        val e1 = EvaluatedMove(Move.byCode("b2-b4"), Ongoing(.1))
        val e2 = EvaluatedMove(Move.byCode("c2-c4"), Ongoing(.1))
        val e3 = EvaluatedMove(Move.byCode("b2-b4"), Ongoing(-.1))
        val e4 = EvaluatedMove(Move.byCode("f2-f4"), Ongoing(1.2))

        assertTrue(e1.compareTo(e1) == 0)
        assertTrue(e1.compareTo(e2) == 0)
        assertTrue(e1 > e3)
        assertTrue(e1 < e4)
    }

    @Test
    fun testOrderingInSortedSet() {

        val set = TreeSet<EvaluatedMove>()

        set.add(EvaluatedMove(Move.byCode("a2-a3"), Ongoing(-1.2)))//7.
        set.add(EvaluatedMove(Move.byCode("b2-b3"), Ongoing(-0.5)))//6.
        set.add(EvaluatedMove(Move.byCode("c2-c3"), Ongoing(4.6)))//3.
        set.add(EvaluatedMove(Move.byCode("d2-d3"), Ongoing(2.1)))//4.
        set.add(EvaluatedMove(Move.byCode("e2-e3"), Draw))//5.
        set.add(EvaluatedMove(Move.byCode("f2-f3"), CheckmateSelf(1)))//9.
        set.add(EvaluatedMove(Move.byCode("g2-g3"), CheckmateSelf(2)))//8.
        set.add(EvaluatedMove(Move.byCode("h2-h3"), CheckmateOther(1)))//1.
        set.add(EvaluatedMove(Move.byCode("h2-h4"), CheckmateOther(2)))//2.

        val iter = set.descendingIterator()

        assertEquals(iter.next().move, Move.byCode("h2-h3"))
        assertEquals(iter.next().move, Move.byCode("h2-h4"))
        assertEquals(iter.next().move, Move.byCode("c2-c3"))
        assertEquals(iter.next().move, Move.byCode("d2-d3"))
        assertEquals(iter.next().move, Move.byCode("e2-e3"))
        assertEquals(iter.next().move, Move.byCode("b2-b3"))
        assertEquals(iter.next().move, Move.byCode("a2-a3"))
        assertEquals(iter.next().move, Move.byCode("g2-g3"))
        assertEquals(iter.next().move, Move.byCode("f2-f3"))
    }
}
